src/rt/rust_builtin.c

```diff
@@ -199,6 +199,56 @@ rust_unset_sigprocmask() {
 int *__dfly_error(void) { return __error(); }
 #endif
 
+#if defined(__OpenBSD__)
+#include <sys/param.h>
+#include <sys/sysctl.h>
+#include <limits.h>
+
+const char * __load_self() {
+    static char *self = NULL;
+
+    if (self == NULL) {
+        int mib[4];
+        char **argv = NULL;
+        size_t argv_len;
+
+        /* initialize mib */
+        mib[0] = CTL_KERN;
+        mib[1] = KERN_PROC_ARGS;
+        mib[2] = getpid();
+        mib[3] = KERN_PROC_ARGV;
+
+        /* request KERN_PROC_ARGV size */
+        argv_len = 0;
+        if (sysctl(mib, 4, NULL, &argv_len, NULL, 0) == -1)
+            return (NULL);
+
+        /* allocate size */
+        if ((argv = malloc(argv_len)) == NULL)
+            return (NULL);
+
+        /* request KERN_PROC_ARGV */
+        if (sysctl(mib, 4, argv, &argv_len, NULL, 0) == -1) {
+            free(argv);
+            return (NULL);
+        }
+
+        /* get realpath if possible */
+        if ((argv[0] != NULL) && ((*argv[0] == '.') || (*argv[0] == '/')
+				|| (strstr(argv[0], "/") != NULL)))
+
+            self = realpath(argv[0], NULL);
```

Comment:
I am not sure to understand the purpose of the question... if it is about the whole function (we try to initialise `self`, and return it only if possible), or about the `if` before the call of `realpath`.

argv[0] could have several kind of values, depending how the executable was called.

```
$ ./my-binary
argv[0] = "./my-binary"

$ /tmp/my-binary
argv[0] = "/tmp/my-binary"

$ cd / ; tmp/my-binary
argv[0] = "tmp/my-binary"

$ PATH=/tmp my-binary
argv[0] = "my-binary"
```

The 3 first case are "safe": the `realpath(argv[0], NULL)` will return the filename (and if there is no race [file deleted or replaced between launch and call], it should be correct).

But for the last case (binary in PATH), `realpath` couldn't do the trick. The return of `realpath` depend of `cwd`, and when `PATH` is used there is no link between `cwd` and `argv[0]`.

It is a limitation of the implementation of load_self under OpenBSD. The reason OpenBSD don't provide a manner to obtain the filepath of a running executable should be the race exposed before: there is none safe manner to ensure that the filename that would be provided would be the running executable.

For rustc, load_self is used for set a default value to sysroot. The provided implementation permit that under the build process. But when installed (and used in PATH) the problem is here. But it is a packaging problem and not need to be resolved upstream (that could resolved by providing a shell wrapper, or by having a "/usr/local" default as fallback).